File: C++/robot_return_to_origin.cpp
#include <iostream>
#include "./robot_return_to_origin.h"

using namespace std;

bool Solution::judgeCircle(string moves) {
    Cordinate c;
    // O(moves.length())
    for (int i = 0; i < moves.length(); ++i) {
        switch (moves[i]) {
            case 'D':
                movedown(c);
                break;
            case 'U':
                moveup(c);
                break;
            case 'L':
                moveleft(c);
                break;
            case 'R':
                moveright(c);
                break;
            default:
                cout << "unknown move\n";
                return false;
        }
    }
    if (c.x == 0 && c.y == 0) return true;
    return false;
}

int main() {
    string s = "UD";

    Solution* sol = new Solution();
    bool output = sol->judgeCircle(s);
    delete sol;

    cout << output << "\n";

    return 0;
}

File: Python2/robot_return_to_origin.py
class Solution(object):
    def judgeCircle(self, s):
        cordinate = [0, 0]
        for direction in s:
            if direction == 'U':
                cordinate[0] += 1
            elif direction == 'D':
                cordinate[0] -= 1
            elif direction == 'L':
                cordinate[1] -= 1
            elif direction == 'R':
                cordinate[1] += 1
            else:
                print 'Unknown move'
                return False
        if cordinate[0] == 0 and cordinate[1] == 0:
            return True
        return False

    def judgeCircleByCount(self, s):
        # count() is O(n), but it is implemented in native code.
        # So it will be faster than normal Python implementation
        # And we just have to make sure the number of L/R and U/D are the same
        return s.count('L') == s.count('R') and s.count('U') == s.count('D')


if __name__ == '__main__':
    s = 'RRDD'
    sol = Solution()
    print sol.judgeCircleByCount(s)
